mode_calculations.py

```diff
@@ -600,3 +601,366 @@ def inner_product(t, abar, b, axis=None, apply_conjugate=False):
+        Resulting tuple has length two.  The first element is the set
+        of times u for the timeseries.  The second element is the
+        timeseries of <a|M|b>(u).
+
+    """
+
+    from .extrapolation import intersection
+    from spherical_functions import LM_index
+
+    if (a.spin_weight != b.spin_weight):
+        raise ValueError("Spin weights must match in matrix_expect_val")
+
+    LM_clip = slice(a.ell_min, a.ell_max + 1)
+    if ((a.ell_min != b.ell_min) or (a.ell_max != b.ell_max)):
+        if (allow_LM_differ):
+            LM_clip = slice( max(a.ell_min, b.ell_min),
+                             min(a.ell_max, b.ell_max) + 1 )
+            if (LM_clip.start >= LM_clip.stop):
+                raise ValueError("Intersection of (ell,m) modes is "
+                                 "empty.  Assuming this is not desired.")
+        else:
+            raise ValueError("ell_min and ell_max must match in matrix_expect_val "
+                             "(use allow_LM_differ=True to override)")
+
+    t_clip = None
+    if not np.array_equal(a.t, b.t):
+        if (allow_times_differ):
+            t_clip = intersection(a.t, b.t)
+        else:
+            raise ValueError("Time samples must match in matrix_expect_val "
+                             "(use allow_times_differ=True to override)")
+
+    ##########
+
+    times = a.t
+    A = a
+    B = b
+
+    if (LM_clip is not None):
+        A = A[:,LM_clip]
+        B = B[:,LM_clip]
+
+    if (t_clip is not None):
+        times = t_clip
+        A = A.interpolate(t_clip)
+        B = B.interpolate(t_clip)
+
+    ##########
+    ## Time for actual math!
+
+    Abar_data  = np.conj(A.data)
+    B_data     = B.data
+
+    expect_val = np.zeros(times.size, dtype=np.complex)
+
+    ell_min = LM_clip.start
+    ell_max = LM_clip.stop - 1
+
+    for ellp, mp, ell, m, M_el in M(ell_min, ell_max):
+        # QUESTION: Should we test if M_el is very close to 0, and
+        # continue to the next element if it's close enough?
+
+        i_a = LM_index(ellp, mp, A.ell_min)
+        i_b = LM_index(ell , m , B.ell_min)
+
+        # QUESTION: Do any python implementations short-circuit the
+        # multiplication if M_el is 0?
+        expect_val += M_el * Abar_data[:, i_a] * B_data[:, i_b]
+
+    return times, expect_val
+
+def id_M(ell_min, ell_max):
+    """Generator for identity matrix elements (for use with matrix_expect_val)
+
+    Matrix elements yielded are
+    <ellp, mp|I|ell, m> = \delta_{ellp ell} \delta_{mp, m}
+    """
+
+    for ell in xrange(ell_min, ell_max+1):
+        for m in xrange(-ell, ell+1):
+            yield ell, m, ell, m, 1.
+
+def j_z_M(ell_min, ell_max):
+    """Generator for j^z matrix elements (for use with matrix_expect_val)
+
+    Matrix elements yielded are
+    <ellp, mp|j^z|ell, m> = 1.j * m \delta_{ellp ell} \delta_{mp, m}
+    This follows the convention for j^z from Ruiz+ (2008) [0707.4654]
+    """
+
+    for ell in xrange(ell_min, ell_max+1):
```

Comment:
FYI, `xrange` is a holdover from python 2.  Since `scri` is now evidently just a python 3 package, you should feel free to just use `range`.

Reply:
Done (for the code I added, I did not touch instances of `xrange` in other places)